common/testing/testlogger/testlogger.go

```diff
@@ -0,0 +1,504 @@
+		tl.wrapped.Info(fmt.Sprintf("(%p) TestLogger::NotExpecting: '%s'\n", tl, msg))
+	}
+	return tl.addExpectationLocked(true, level, msg, tags...)
+}
+
+func (tl *TestLogger) addExpectationLocked(unexpected bool, level Level, msg string, tags ...tag.Tag) *Expectation {
+	e := &Expectation{
+		testLogger: tl,
+		lvl:        level,
+		unexpected: unexpected,
+	}
+	m := newMatcher(msg, tags, e)
+	e.e = tl.state.mu.expectations[level].PushBack(m)
+	return e
+}
+
+// Forget removes a previously registered expectation.
+// It will no longer be used to determine if a log message is expected or unexpected.
+func (tl *TestLogger) Forget(e *Expectation) {
+	tl.state.mu.Lock()
+	defer tl.state.mu.Unlock()
+	tl.state.mu.expectations[e.lvl].Remove(e.e)
+}
+
+func (tl *TestLogger) isUnexpected(level Level, msg string, tags []tag.Tag) bool {
+	expectations := tl.state.mu.expectations[level]
+	for e := expectations.Front(); e != nil; e = e.Next() {
+		pat, ok := e.Value.(matcher)
+		if !ok {
+			tl.state.t.Fatalf("Bug in TestLogger: invalid %T value in matcher list", e.Value)
+		}
+		if pat.Matches(msg, tags) {
+			return pat.expectation.unexpected
+		}
+	}
+	return false
+}
+
+// FailOnError overrides the behavior of this logger. It returns the previous value
+// so that it can be restored later.
+func (tl *TestLogger) FailOnError(v bool) bool {
+	return tl.state.failOnError.Swap(v)
+}
+
+// DPanic implements log.Logger.
+func (tl *TestLogger) DPanic(msg string, tags ...tag.Tag) {
+	tl.state.mu.RLock()
+	defer tl.state.mu.RUnlock()
+	if tl.state.mu.closed {
+		return
+	}
+	if tl.tags != nil {
+		tags = append(tags, tl.tags...)
+	}
+	// note, actual panic'ing in wrapped is turned off so we can control.
+	tl.wrapped.DPanic(msg, tags...)
+	if tl.state.failOnDPanic.Load() && tl.isUnexpected(Error, msg, tags) {
+		tl.state.t.Helper()
+		panic(failureMessage("DPanic", msg, tags))
+	}
+}
+
+// Debug implements log.Logger.
+func (tl *TestLogger) Debug(msg string, tags ...tag.Tag) {
+	tl.state.mu.RLock()
+	defer tl.state.mu.RUnlock()
+	if tl.state.mu.closed {
+		return
+	}
+	if tl.tags != nil {
+		tags = append(tags, tl.tags...)
+	}
+	_ = tl.isUnexpected(Debug, msg, tags)
+	tl.wrapped.Debug(msg, append(tags, tl.tags...)...)
+}
+
+// Error implements log.Logger.
+func (tl *TestLogger) Error(msg string, tags ...tag.Tag) {
+	tl.state.mu.RLock()
+	if tl.state.mu.closed {
+		tl.state.mu.RUnlock()
+		return
+	}
+	if tl.tags != nil {
+		tags = append(tags, tl.tags...)
+	}
+	if !tl.isUnexpected(Error, msg, tags) {
+		tl.wrapped.Error(msg, tags...)
+		tl.state.mu.RUnlock()
+		return
+	}
+	tl.state.mu.RUnlock()
+
+	if tl.state.failOnError.Load() {
+		tl.state.t.Helper()
+		tl.wrapped.Error(msg, tags...)
+		panic(failureMessage("Error", msg, tags))
+	}
+
+	// Labeling the error as unexpected; so it can be found later.
+	tl.wrapped.Error(errorMessage("Error", msg), tags...)
+	tl.state.mu.Lock()
+	tl.state.mu.unexpectedErrs = true
+	tl.state.mu.Unlock()
```

Comment:
Still logs the error, but adds a prefix to make it possible to find the error in the logs.

Reply:
Why are you setting `unexpectedErrs` here if it wasn't already set?

Reply:
I'm sorry, I don't understand the question.